Matrix operations

tudny opened the pull request.

tudny committed:
WIP matrix ops
added add
added sub
wip neg

tudny committed:
No idea how to fix it

tudny committed:
Working neg

mgr0dzicki committed:
schodki schodki

mgr0dzicki commented on src/matrices.rs:
```diff
@@ -17,8 +22,191 @@ impl<T: MatrixNumber> Matrix<T> {
+
+    fn check_shape(&self, other: &Self) -> anyhow::Result<()> {
+        if self.data.len() != other.data.len() {
+            return Err(anyhow::anyhow!("Matrices have different number of rows!"));
```
`return Err(anyhow::anyhow!(...));` $\equiv$ `anyhow::bail!(...)`

tudny committed:
Fixed anyhow

return Err(anyhow::anyhow!(...)); $\equiv$ anyhow::bail!(...)

`cargo fmt` + `cargo clippy`

tudny committed:
dirty test commit

tudny committed:
undo: dirty test commit

tudny committed:
Started Mul

tudny committed:
Merge remote-tracking branch 'origin/matrix-ops' into matrix-ops

# Conflicts:
#	src/matrices.rs
#	src/traits.rs

tudny commented:
This was the ugliest conflict I had to deal with

tudny committed:
for loops need to be trashed later

tudny committed:
fixed clippy

tudny committed:
fixed fmt

tudny committed:
Added macros for rational and normal matrices

tudny commented on src/matrices.rs:
```diff
@@ -40,24 +248,280 @@ impl<T: MatrixNumber> LaTeXable for Matrix<T> {
+    }
+}
+
+#[macro_export]
+macro_rules! rv {
+    ($($x:expr),+ $(,)?) => (
+        vec![
+            $(ri!($x)),+
+        ]
+    );
+}
+
+#[macro_export]
+macro_rules! rm {
+    ($($($x:expr),+ $(,)?);+ $(;)?) => (
+        Matrix::<Rational64>::new(vec![
+            $(rv!($($x),+)),+
+        ])
+    );
+}
+
+#[macro_export]
+macro_rules! im {
+    ($($($x:expr),+ $(,)?);+ $(;)?) => (
+        Matrix::new(vec![
+            $(vec![
+                $($x),+
+            ]),+
+        ])
+    );
+}
```
@mgr0dzicki check out those wild macros!

tudny committed:
macro rules are now safe (in runtime)

tudny committed:
refactor

tudny committed:
refactor 2

tudny committed:
Update matrices.rs

tudny committed:
default implementation for to_latex_single

mgr0dzicki commented on src/main.rs:
```diff
@@ -1,3 +1,5 @@
+extern crate core;
```
why

tudny replied:
who asked

mgr0dzicki committed:
remove extern crate core

mgr0dzicki committed:
Merge branch 'matrix-ops' of https://github.com/mimuw-jnp2-rust/project-matrix-busters into matrix-ops

tudny committed:
review commit

mgr0dzicki reviewed (approved):
niech ci będzie
